File: frontend/nextjs-app/src/lib/services/connectivity-service.ts
/**
 * Connectivity service for monitoring network status in Next.js
 */
class ConnectivityService {
  private listeners: Set<(status: ConnectivityStatus) => void> = new Set();
  private currentStatus: ConnectivityStatus = ConnectivityStatus.UNKNOWN;

  constructor() {
    if (typeof window !== 'undefined') {
      this.init();
    }
  }

  /**
   * Initialize connectivity monitoring
   */
  private init(): void {
    // Set initial status
    this.updateStatus();

    // Listen to online/offline events
    window.addEventListener('online', () => {
      console.log('📡 Network: Online');
      this.updateStatus();
    });

    window.addEventListener('offline', () => {
      console.log('📡 Network: Offline');
      this.updateStatus();
    });

    // Listen to connection change events (if supported)
    if ('connection' in navigator) {
      const connection = (navigator as any).connection;
      connection?.addEventListener('change', () => {
        console.log('📡 Network: Connection changed');
        this.updateStatus();
      });
    }

    console.log('📡 Connectivity service initialized');
  }

  /**
   * Update connectivity status
   */
  private updateStatus(): void {
    const newStatus = navigator.onLine 
      ? ConnectivityStatus.CONNECTED 
      : ConnectivityStatus.DISCONNECTED;

    if (this.currentStatus !== newStatus) {
      this.currentStatus = newStatus;
      this.notifyListeners(newStatus);
    }
  }

  /**
   * Notify all listeners of status change
   */
  private notifyListeners(status: ConnectivityStatus): void {
    this.listeners.forEach(listener => {
      try {
        listener(status);
      } catch (error) {
        console.error('❌ Error notifying connectivity listener:', error);
      }
    });
  }

  /**
   * Get current connectivity status
   */
  getCurrentStatus(): ConnectivityStatus {
    return this.currentStatus;
  }

  /**
   * Check if device is connected to internet
   */
  async isConnected(): Promise<boolean> {
    if (typeof window === 'undefined') return true; // Assume connected on server
    
    // Quick check using navigator.onLine
    if (!navigator.onLine) {
      return false;
    }

    // Additional check by trying to fetch a small resource
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 3000);

      await fetch('/api/health', {
        method: 'HEAD',
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Subscribe to connectivity changes
   */
  subscribe(listener: (status: ConnectivityStatus) => void): () => void {
    this.listeners.add(listener);

    // Return unsubscribe function
    return () => {
      this.listeners.delete(listener);
    };
  }

  /**
   * Get connection type (if available)
   */
  getConnectionType(): string {
    if (typeof window === 'undefined') return 'unknown';

    const connection = (navigator as any).connection;
    if (connection) {
      return connection.effectiveType || connection.type || 'unknown';
    }

    return 'unknown';
  }

  /**
   * Check if connection is slow
   */
  isSlowConnection(): boolean {
    if (typeof window === 'undefined') return false;

    const connection = (navigator as any).connection;
    if (connection) {
      const effectiveType = connection.effectiveType;
      return effectiveType === 'slow-2g' || effectiveType === '2g';
    }

    return false;
  }

  /**
   * Get estimated downlink speed (Mbps)
   */
  getDownlinkSpeed(): number | null {
    if (typeof window === 'undefined') return null;

    const connection = (navigator as any).connection;
    return connection?.downlink || null;
  }
}

/**
 * Connectivity status enum
 */
export enum ConnectivityStatus {
  CONNECTED = 'connected',
  DISCONNECTED = 'disconnected',
  UNKNOWN = 'unknown',
}

/**
 * Helper functions for connectivity status
 */
export const connectivityHelpers = {
  isConnected: (status: ConnectivityStatus) => status === ConnectivityStatus.CONNECTED,
  isDisconnected: (status: ConnectivityStatus) => status === ConnectivityStatus.DISCONNECTED,
  
  getMessage: (status: ConnectivityStatus): string => {
    switch (status) {
      case ConnectivityStatus.CONNECTED:
        return 'Connected to internet';
      case ConnectivityStatus.DISCONNECTED:
        return 'No internet connection';
      case ConnectivityStatus.UNKNOWN:
        return 'Checking connection...';
    }
  },

  getIcon: (status: ConnectivityStatus): string => {
    switch (status) {
      case ConnectivityStatus.CONNECTED:
        return '✅';
      case ConnectivityStatus.DISCONNECTED:
        return '❌';
      case ConnectivityStatus.UNKNOWN:
        return '❓';
    }
  },
};

// Export singleton instance
export const connectivityService = new ConnectivityService();

File: frontend/mobile/lib/core/services/connectivity_service.dart
import 'dart:async';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/foundation.dart';

/// Service for monitoring network connectivity
class ConnectivityService {
  static final ConnectivityService _instance = ConnectivityService._internal();
  factory ConnectivityService() => _instance;
  ConnectivityService._internal();
  
  final Connectivity _connectivity = Connectivity();
  
  StreamController<ConnectivityStatus>? _connectivityController;
  StreamSubscription<List<ConnectivityResult>>? _connectivitySubscription;
  
  ConnectivityStatus _currentStatus = ConnectivityStatus.unknown;
  
  /// Get current connectivity status
  ConnectivityStatus get currentStatus => _currentStatus;
  
  /// Stream of connectivity status changes
  Stream<ConnectivityStatus> get onConnectivityChanged {
    _connectivityController ??= StreamController<ConnectivityStatus>.broadcast();
    return _connectivityController!.stream;
  }
  
  /// Initialize connectivity monitoring
  Future<void> init() async {
    // Get initial connectivity status
    await _updateConnectivityStatus();
    
    // Listen to connectivity changes
    _connectivitySubscription = _connectivity.onConnectivityChanged.listen(
      (List<ConnectivityResult> results) {
        _handleConnectivityChange(results);
      },
    );
    
    if (kDebugMode) {
      print('📡 Connectivity service initialized');
      print('📡 Current status: ${_currentStatus.name}');
    }
  }
  
  /// Dispose connectivity service
  void dispose() {
    _connectivitySubscription?.cancel();
    _connectivityController?.close();
  }
  
  /// Check if device is connected to internet
  Future<bool> isConnected() async {
    await _updateConnectivityStatus();
    return _currentStatus == ConnectivityStatus.connected;
  }
  
  /// Check if device is connected via WiFi
  Future<bool> isWiFi() async {
    final results = await _connectivity.checkConnectivity();
    return results.contains(ConnectivityResult.wifi);
  }
  
  /// Check if device is connected via mobile data
  Future<bool> isMobile() async {
    final results = await _connectivity.checkConnectivity();
    return results.contains(ConnectivityResult.mobile);
  }
  
  /// Update connectivity status
  Future<void> _updateConnectivityStatus() async {
    try {
      final results = await _connectivity.checkConnectivity();
      _handleConnectivityChange(results);
    } catch (e) {
      if (kDebugMode) {
        print('❌ Error checking connectivity: $e');
      }
      _updateStatus(ConnectivityStatus.unknown);
    }
  }
  
  /// Handle connectivity change
  void _handleConnectivityChange(List<ConnectivityResult> results) {
    if (results.isEmpty || results.contains(ConnectivityResult.none)) {
      _updateStatus(ConnectivityStatus.disconnected);
    } else if (results.contains(ConnectivityResult.wifi)) {
      _updateStatus(ConnectivityStatus.connected);
    } else if (results.contains(ConnectivityResult.mobile)) {
      _updateStatus(ConnectivityStatus.connected);
    } else if (results.contains(ConnectivityResult.ethernet)) {
      _updateStatus(ConnectivityStatus.connected);
    } else {
      _updateStatus(ConnectivityStatus.unknown);
    }
  }
  
  /// Update status and notify listeners
  void _updateStatus(ConnectivityStatus newStatus) {
    if (_currentStatus != newStatus) {
      final oldStatus = _currentStatus;
      _currentStatus = newStatus;
      
      if (kDebugMode) {
        print('📡 Connectivity changed: ${oldStatus.name} → ${newStatus.name}');
      }
      
      _connectivityController?.add(newStatus);
    }
  }
}

/// Connectivity status enum
enum ConnectivityStatus {
  /// Device is connected to internet
  connected,
  
  /// Device is not connected to internet
  disconnected,
  
  /// Connectivity status is unknown
  unknown,
}

/// Extension for connectivity status
extension ConnectivityStatusExtension on ConnectivityStatus {
  /// Check if connected
  bool get isConnected => this == ConnectivityStatus.connected;
  
  /// Check if disconnected
  bool get isDisconnected => this == ConnectivityStatus.disconnected;
  
  /// Get user-friendly message
  String get message {
    switch (this) {
      case ConnectivityStatus.connected:
        return 'Connected to internet';
      case ConnectivityStatus.disconnected:
        return 'No internet connection';
      case ConnectivityStatus.unknown:
        return 'Checking connection...';
    }
  }
  
  /// Get icon for status
  String get icon {
    switch (this) {
      case ConnectivityStatus.connected:
        return '✅';
      case ConnectivityStatus.disconnected:
        return '❌';
      case ConnectivityStatus.unknown:
        return '❓';
    }
  }
}
